// @resonatehq/sdk/dist/sim/example-3.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || (function () {
    var ownKeys = function(o) {
        ownKeys = Object.getOwnPropertyNames || function (o) {
            var ar = [];
            for (var k in o) if (Object.prototype.hasOwnProperty.call(o, k)) ar[ar.length] = k;
            return ar;
        };
        return ownKeys(o);
    };
    return function (mod) {
        if (mod && mod.__esModule) return mod;
        var result = {};
        if (mod != null) for (var k = ownKeys(mod), i = 0; i < k.length; i++) if (k[i] !== "default") __createBinding(result, mod, k[i]);
        __setModuleDefault(result, mod);
        return result;
    };
})();
Object.defineProperty(exports, "__esModule", { value: true });
const clock_1 = require("../src/clock");
const encoder_1 = require("../src/encoder");
const registry_1 = require("../src/registry");
const util = __importStar(require("../src/util"));
const server_1 = require("./src/server");
const simulator_1 = require("./src/simulator");
const worker_1 = require("./src/worker");
// Function definition
function* fibonacci(ctx, n) {
    if (n <= 1) {
        return n;
    }
    const p1 = yield ctx.beginRpc("fibonacci", n - 1, ctx.options({ id: `fibonacci-${n - 1}` }));
    const p2 = yield ctx.beginRpc("fibonacci", n - 2, ctx.options({ id: `fibonacci-${n - 2}` }));
    return (yield p1) + (yield p2);
}
const options = { seed: 0, steps: 100, randomDelay: 0, dropProb: 0, duplProb: 0, charFlipProb: 0 };
// Run Simulation
const rnd = new simulator_1.Random(options.seed);
const clock = new clock_1.StepClock();
const encoder = new encoder_1.JsonEncoder();
const registry = new registry_1.Registry();
registry.add(fibonacci);
const sim = new simulator_1.Simulator(rnd, {
    randomDelay: options.randomDelay,
    dropProb: options.dropProb,
    duplProb: options.duplProb,
});
const server = new server_1.ServerProcess(clock, "server");
const worker1 = new worker_1.WorkerProcess(rnd, clock, encoder, registry, { charFlipProb: options.charFlipProb ?? rnd.random(0.05) }, "worker-1", "default");
const worker2 = new worker_1.WorkerProcess(rnd, clock, encoder, registry, { charFlipProb: options.charFlipProb ?? rnd.random(0.05) }, "worker-2", "default");
const worker3 = new worker_1.WorkerProcess(rnd, clock, encoder, registry, { charFlipProb: options.charFlipProb ?? rnd.random(0.05) }, "worker-3", "default");
const workers = [worker1, worker2, worker3];
sim.register(server);
for (const worker of workers) {
    sim.register(worker);
}
const n = 10;
const id = `fibonacci-${n}`;
sim.delay(0, () => {
    const msg = new simulator_1.Message((0, simulator_1.unicast)("environment"), (0, simulator_1.unicast)("server"), {
        kind: "createPromise",
        id,
        timeout: 10000000000,
        iKey: id,
        tags: { "resonate:invoke": "local://any@default" },
        param: encoder.encode({ func: "fibonacci", args: [n], version: 1 }),
    }, { requ: true, correlationId: 1 });
    sim.send(msg);
});
sim.exec(options.steps);
function f(n, memo = {}) {
    if (n <= 1)
        return n;
    if (memo[n] !== undefined) {
        return memo[n];
    }
    memo[n] = f(n - 1, memo) + f(n - 2, memo);
    return memo[n];
}
util.assert(encoder.decode(server.server.promises.get(id)?.value) === f(n));
