// @resonatehq/sdk/dist/dev/network.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || (function () {
    var ownKeys = function(o) {
        ownKeys = Object.getOwnPropertyNames || function (o) {
            var ar = [];
            for (var k in o) if (Object.prototype.hasOwnProperty.call(o, k)) ar[ar.length] = k;
            return ar;
        };
        return ownKeys(o);
    };
    return function (mod) {
        if (mod && mod.__esModule) return mod;
        var result = {};
        if (mod != null) for (var k = ownKeys(mod), i = 0; i < k.length; i++) if (k[i] !== "default") __createBinding(result, mod, k[i]);
        __setModuleDefault(result, mod);
        return result;
    };
})();
Object.defineProperty(exports, "__esModule", { value: true });
exports.LocalMessageSource = exports.LocalNetwork = void 0;
const util = __importStar(require("../src/util"));
const server_1 = require("./server");
class LocalNetwork {
    server;
    messageSource;
    constructor({ pid = "pid", group = "default", server = new server_1.Server(), } = {}) {
        this.server = server;
        this.messageSource = new LocalMessageSource(pid, group, server);
    }
    getMessageSource() {
        return this.messageSource;
    }
    start() { }
    stop() { }
    send(req, callback) {
        setTimeout(() => {
            try {
                const res = this.server.process(req, Date.now());
                util.assert(res.kind === req.kind, "res kind must match req kind");
                callback(undefined, res);
                this.messageSource.enqueueNext();
            }
            catch (err) {
                callback(err);
            }
        });
    }
}
exports.LocalNetwork = LocalNetwork;
class LocalMessageSource {
    pid;
    group;
    unicast;
    anycast;
    server;
    timeoutId;
    shouldStop = false;
    subscriptions = { invoke: [], resume: [], notify: [] };
    constructor(pid, group, server) {
        this.pid = pid;
        this.group = group;
        this.unicast = `poll://uni@${group}/${pid}`;
        this.anycast = `poll://any@${group}/${pid}`;
        this.server = server;
        this.timeoutId = undefined;
    }
    start() { }
    stop() {
        clearTimeout(this.timeoutId);
        this.shouldStop = true;
        this.timeoutId = undefined;
    }
    recv(msg) {
        for (const callback of this.subscriptions[msg.type]) {
            callback(msg);
        }
    }
    subscribe(type, callback) {
        this.subscriptions[type].push(callback);
    }
    enqueueNext() {
        clearTimeout(this.timeoutId);
        const time = Date.now();
        const next = this.server.next(time);
        if (next !== undefined && !this.shouldStop) {
            this.timeoutId = setTimeout(() => {
                for (const { msg } of this.server.step(time)) {
                    this.recv(msg);
                }
                this.enqueueNext();
            }, next);
        }
    }
    match(target) {
        return `poll://any@${target}`;
    }
}
exports.LocalMessageSource = LocalMessageSource;
