// @resonatehq/sdk/dist/src/encoder.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || (function () {
    var ownKeys = function(o) {
        ownKeys = Object.getOwnPropertyNames || function (o) {
            var ar = [];
            for (var k in o) if (Object.prototype.hasOwnProperty.call(o, k)) ar[ar.length] = k;
            return ar;
        };
        return ownKeys(o);
    };
    return function (mod) {
        if (mod && mod.__esModule) return mod;
        var result = {};
        if (mod != null) for (var k = ownKeys(mod), i = 0; i < k.length; i++) if (k[i] !== "default") __createBinding(result, mod, k[i]);
        __setModuleDefault(result, mod);
        return result;
    };
})();
Object.defineProperty(exports, "__esModule", { value: true });
exports.JsonEncoder = void 0;
const util = __importStar(require("./util"));
class JsonEncoder {
    inf = "__INF__";
    negInf = "__NEG_INF__";
    encode(value) {
        // note about undefined:
        // undefined is not json serializable, so immediately return undefined
        if (value === undefined) {
            return { data: undefined };
        }
        const json = JSON.stringify(value, (_, v) => {
            if (v === Number.POSITIVE_INFINITY)
                return this.inf;
            if (v === Number.NEGATIVE_INFINITY)
                return this.negInf;
            if (v instanceof AggregateError) {
                return {
                    __type: "aggregate_error",
                    message: v.message,
                    stack: v.stack,
                    name: v.name,
                    errors: v.errors,
                };
            }
            if (v instanceof Error) {
                return {
                    __type: "error",
                    message: v.message,
                    stack: v.stack,
                    name: v.name,
                };
            }
            return v;
        });
        return {
            headers: {},
            data: util.base64Encode(json),
        };
    }
    decode(value) {
        if (!value?.data) {
            return undefined;
        }
        return JSON.parse(util.base64Decode(value.data), (_, v) => {
            if (v === this.inf)
                return Number.POSITIVE_INFINITY;
            if (v === this.negInf)
                return Number.NEGATIVE_INFINITY;
            if (v?.__type === "aggregate_error") {
                return Object.assign(new AggregateError(v.errors, v.message), v);
            }
            if (v?.__type === "error") {
                const err = new Error(v.message || "Unknown error");
                if (v.name)
                    err.name = v.name;
                if (v.stack)
                    err.stack = v.stack;
                return err;
            }
            return v;
        });
    }
}
exports.JsonEncoder = JsonEncoder;
