// @resonatehq/sdk/dist/src/exceptions.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.ResonateError = void 0;
class ResonateError extends Error {
    code;
    type;
    next;
    href;
    retriable;
    serverError;
    constructor(code, type, mesg, { next = "n/a", cause, retriable = false, serverError, } = {}) {
        super(mesg, { cause });
        this.name = "ResonateError";
        this.code = code;
        this.type = type;
        this.next = next;
        this.href = `https://rn8.io/e/11${code}`; // 11 is the typescript sdk code
        this.retriable = retriable;
        this.serverError = serverError;
    }
    log(verbose) {
        console.error(`${this.type}. ${this.message}. ${this.next}. (See ${this.href} for more information)`);
        if (verbose) {
            console.error(this);
        }
    }
}
exports.ResonateError = ResonateError;
exports.default = {
    REGISTRY_VERSION_INVALID: (v) => {
        return new ResonateError("00", "Registry", `Function version must be greater than zero (${v} provided)`);
    },
    REGISTRY_NAME_REQUIRED: () => {
        return new ResonateError("01", "Registry", "Function name is required");
    },
    REGISTRY_FUNCTION_ALREADY_REGISTERED: (f, v, u) => {
        const under = u ? ` under '${u}'` : "";
        return new ResonateError("02", "Registry", `Function '${f}' (version ${v}) is already registered${under}`);
    },
    REGISTRY_FUNCTION_NOT_REGISTERED: (f, v) => {
        const version = v > 0 ? ` (version ${v})` : "";
        return new ResonateError("03", "Registry", `Function '${f}'${version} is not registered`, {
            next: "Will drop",
        });
    },
    DEPENDENCY_ALREADY_REGISTERED: (d) => {
        return new ResonateError("04", "Dependencies", `Dependency '${d}' is already registered`);
    },
    DEPENDENCY_NOT_REGISTERED: (d) => {
        return new ResonateError("05", "Dependencies", `Dependency '${d}' is not registered`, {
            next: "Will drop",
        });
    },
    ENCODING_ARGS_UNENCODEABLE: (f, c) => {
        return new ResonateError("06", "Encoding", `Argument(s) for function '${f}' cannot be encoded`, {
            next: "Will drop",
            cause: c,
        });
    },
    ENCODING_ARGS_UNDECODEABLE: (f, c) => {
        return new ResonateError("07", "Encoding", `Argument(s) for function '${f}' cannot be decoded`, {
            next: "Will drop",
            cause: c,
        });
    },
    ENCODING_RETV_UNENCODEABLE: (f, c) => {
        return new ResonateError("08", "Encoding", `Return value from function '${f}' cannot be encoded`, {
            next: "Will drop",
            cause: c,
        });
    },
    ENCODING_RETV_UNDECODEABLE: (f, c) => {
        return new ResonateError("09", "Encoding", `Return value from function '${f}' cannot be decoded`, {
            next: "Will drop",
            cause: c,
        });
    },
    PANIC: (src, msg) => {
        src = src.charAt(0).toUpperCase() + src.slice(1);
        msg = msg ? `: ${msg}` : "";
        return new ResonateError("98", "Panic", `${src}${msg}`, {
            next: "Will drop",
        });
    },
    SERVER_ERROR: (m, r, e) => {
        return new ResonateError("99", "Server", m, { retriable: r, serverError: e });
    },
};
