// @resonatehq/sdk/dist/src/heartbeat.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || (function () {
    var ownKeys = function(o) {
        ownKeys = Object.getOwnPropertyNames || function (o) {
            var ar = [];
            for (var k in o) if (Object.prototype.hasOwnProperty.call(o, k)) ar[ar.length] = k;
            return ar;
        };
        return ownKeys(o);
    };
    return function (mod) {
        if (mod && mod.__esModule) return mod;
        var result = {};
        if (mod != null) for (var k = ownKeys(mod), i = 0; i < k.length; i++) if (k[i] !== "default") __createBinding(result, mod, k[i]);
        __setModuleDefault(result, mod);
        return result;
    };
})();
Object.defineProperty(exports, "__esModule", { value: true });
exports.NoopHeartbeat = exports.AsyncHeartbeat = void 0;
const util = __importStar(require("./util"));
class AsyncHeartbeat {
    network;
    intervalId;
    pid;
    counter = 0;
    delay;
    constructor(pid, delay, network) {
        this.pid = pid;
        this.delay = delay;
        this.network = network;
    }
    start() {
        this.counter++;
        if (!this.intervalId) {
            this.heartbeat();
        }
    }
    heartbeat() {
        this.intervalId = setInterval(() => {
            const counter = this.counter;
            this.network.send({
                kind: "heartbeatTasks",
                processId: this.pid,
            }, (err, res) => {
                if (err)
                    return;
                util.assertDefined(res);
                if (res.tasksAffected === 0) {
                    this.clearIntervalIfMatch(counter);
                }
            });
        }, this.delay);
    }
    stop() {
        this.clearIntervalIfMatch(this.counter);
    }
    clearIntervalIfMatch(counter) {
        if (this.counter === counter) {
            clearInterval(this.intervalId);
            this.intervalId = undefined;
        }
    }
}
exports.AsyncHeartbeat = AsyncHeartbeat;
class NoopHeartbeat {
    start() { }
    stop() { }
}
exports.NoopHeartbeat = NoopHeartbeat;
