// @resonatehq/sdk/dist/src/index.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || (function () {
    var ownKeys = function(o) {
        ownKeys = Object.getOwnPropertyNames || function (o) {
            var ar = [];
            for (var k in o) if (Object.prototype.hasOwnProperty.call(o, k)) ar[ar.length] = k;
            return ar;
        };
        return ownKeys(o);
    };
    return function (mod) {
        if (mod && mod.__esModule) return mod;
        var result = {};
        if (mod != null) for (var k = ownKeys(mod), i = 0; i < k.length; i++) if (k[i] !== "default") __createBinding(result, mod, k[i]);
        __setModuleDefault(result, mod);
        return result;
    };
})();
Object.defineProperty(exports, "__esModule", { value: true });
exports.NoopTracer = exports.ResonateInner = exports.Resonate = exports.Registry = exports.OptionsBuilder = exports.HttpNetwork = exports.NoopHeartbeat = exports.AsyncHeartbeat = exports.Handler = exports.NoopEncryptor = exports.JsonEncoder = exports.core = exports.WallClock = void 0;
var clock_1 = require("./clock");
Object.defineProperty(exports, "WallClock", { enumerable: true, get: function () { return clock_1.WallClock; } });
exports.core = __importStar(require("./core"));
var encoder_1 = require("./encoder");
Object.defineProperty(exports, "JsonEncoder", { enumerable: true, get: function () { return encoder_1.JsonEncoder; } });
var encryptor_1 = require("./encryptor");
Object.defineProperty(exports, "NoopEncryptor", { enumerable: true, get: function () { return encryptor_1.NoopEncryptor; } });
var handler_1 = require("./handler");
Object.defineProperty(exports, "Handler", { enumerable: true, get: function () { return handler_1.Handler; } });
var heartbeat_1 = require("./heartbeat");
Object.defineProperty(exports, "AsyncHeartbeat", { enumerable: true, get: function () { return heartbeat_1.AsyncHeartbeat; } });
Object.defineProperty(exports, "NoopHeartbeat", { enumerable: true, get: function () { return heartbeat_1.NoopHeartbeat; } });
var remote_1 = require("./network/remote");
Object.defineProperty(exports, "HttpNetwork", { enumerable: true, get: function () { return remote_1.HttpNetwork; } });
var options_1 = require("./options");
Object.defineProperty(exports, "OptionsBuilder", { enumerable: true, get: function () { return options_1.OptionsBuilder; } });
var registry_1 = require("./registry");
Object.defineProperty(exports, "Registry", { enumerable: true, get: function () { return registry_1.Registry; } });
var resonate_1 = require("./resonate");
Object.defineProperty(exports, "Resonate", { enumerable: true, get: function () { return resonate_1.Resonate; } });
var resonate_inner_1 = require("./resonate-inner");
Object.defineProperty(exports, "ResonateInner", { enumerable: true, get: function () { return resonate_inner_1.ResonateInner; } });
var tracer_1 = require("./tracer");
Object.defineProperty(exports, "NoopTracer", { enumerable: true, get: function () { return tracer_1.NoopTracer; } });
