// @resonatehq/sdk/dist/src/nursery.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.Nursery = void 0;
class Nursery {
    // event queue, these functions are ensured to execute sequentially
    q = [];
    // the function the nursery is instantiated with, added to the
    // queue when holds are released
    f;
    // the callback the nursery is instantiated with, called once when
    // the nursery is done and all holds are released
    c;
    // the error value done is called with, if any
    e;
    // the return value done is called with, if any
    r;
    holds = 0;
    running = false;
    completed = false;
    constructor(f, c) {
        this.f = () => f(this);
        this.c = c;
        // kick off the nursery
        this.enqueue(this.f);
    }
    hold(f) {
        if (!this.running || this.completed)
            return;
        this.holds++;
        f(() => {
            this.holds--;
            if (this.completed) {
                this.complete();
            }
            else {
                this.enqueue(this.f);
            }
        });
    }
    cont() {
        this.running = false;
        if (this.completed) {
            this.complete();
        }
        else {
            this.execute();
        }
    }
    done(err, res) {
        if (this.completed)
            return;
        this.e = err;
        this.r = res;
        this.running = false;
        this.completed = true;
        this.complete();
    }
    all(list, func, done) {
        const results = new Array(list.length);
        let remaining = list.length;
        let completed = false;
        const finalize = (err) => {
            if (completed)
                return;
            completed = true;
            done(err, results);
        };
        list.forEach((item, index) => {
            func(item, (err, res) => {
                if (completed)
                    return;
                if (err)
                    return finalize(err);
                results[index] = res;
                remaining--;
                if (remaining === 0) {
                    finalize();
                }
            });
        });
    }
    enqueue(f) {
        this.q.push(f);
        this.execute();
    }
    execute() {
        if (this.running)
            return;
        const f = this.q.shift();
        if (f) {
            this.running = true;
            f();
        }
    }
    complete() {
        if (!this.completed || this.holds > 0)
            return;
        this.c(this.e, this.r);
    }
}
exports.Nursery = Nursery;
