// @resonatehq/sdk/dist/src/resonate-inner.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || (function () {
    var ownKeys = function(o) {
        ownKeys = Object.getOwnPropertyNames || function (o) {
            var ar = [];
            for (var k in o) if (Object.prototype.hasOwnProperty.call(o, k)) ar[ar.length] = k;
            return ar;
        };
        return ownKeys(o);
    };
    return function (mod) {
        if (mod && mod.__esModule) return mod;
        var result = {};
        if (mod != null) for (var k = ownKeys(mod), i = 0; i < k.length; i++) if (k[i] !== "default") __createBinding(result, mod, k[i]);
        __setModuleDefault(result, mod);
        return result;
    };
})();
Object.defineProperty(exports, "__esModule", { value: true });
exports.ResonateInner = void 0;
const computation_1 = require("./computation");
const retries_1 = require("./retries");
const util = __importStar(require("./util"));
class ResonateInner {
    unicast;
    anycast;
    pid;
    ttl;
    clock;
    network;
    handler;
    tracer;
    retries;
    registry;
    heartbeat;
    dependencies;
    optsBuilder;
    verbose;
    computations = new Map();
    constructor({ unicast, anycast, pid, ttl, clock, network, handler, tracer, registry, heartbeat, dependencies, optsBuilder, verbose, messageSource = undefined, }) {
        this.unicast = unicast;
        this.anycast = anycast;
        this.pid = pid;
        this.ttl = ttl;
        this.clock = clock;
        this.network = network;
        this.handler = handler;
        this.tracer = tracer;
        this.registry = registry;
        this.heartbeat = heartbeat;
        this.dependencies = dependencies;
        this.optsBuilder = optsBuilder;
        this.verbose = verbose;
        // default retry policies
        this.retries = new Map([
            [retries_1.Constant.type, retries_1.Constant],
            [retries_1.Exponential.type, retries_1.Exponential],
            [retries_1.Linear.type, retries_1.Linear],
            [retries_1.Never.type, retries_1.Never],
        ]);
        // subscribe to invoke and resume
        messageSource?.subscribe("invoke", this.onMessage.bind(this));
        messageSource?.subscribe("resume", this.onMessage.bind(this));
    }
    process(span, task, done) {
        let computation = this.computations.get(task.task.rootPromiseId);
        if (!computation) {
            computation = new computation_1.Computation(task.task.rootPromiseId, this.unicast, this.anycast, this.pid, this.ttl, this.clock, this.network, this.handler, this.retries, this.registry, this.heartbeat, this.dependencies, this.optsBuilder, this.verbose, this.tracer, span);
            this.computations.set(task.task.rootPromiseId, computation);
        }
        computation.process(task, done);
    }
    onMessage(msg) {
        util.assert(msg.type === "invoke" || msg.type === "resume");
        if (msg.type === "invoke" || msg.type === "resume") {
            this.process(this.tracer.decode(msg.headers), { kind: "unclaimed", task: msg.task }, () => { });
        }
    }
}
exports.ResonateInner = ResonateInner;
