// @resonatehq/sdk/dist/src/retries.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.Never = exports.Linear = exports.Exponential = exports.Constant = void 0;
class Constant {
    static type = "constant";
    delay;
    maxRetries;
    constructor({ delay = 1000, maxRetries = Number.MAX_SAFE_INTEGER } = {}) {
        this.delay = delay;
        this.maxRetries = maxRetries;
    }
    next(attempt) {
        if (attempt < 0) {
            throw new Error("attempt must be greater than or equal to 0");
        }
        if (attempt > this.maxRetries) {
            return null;
        }
        if (attempt === 0) {
            return 0;
        }
        return this.delay;
    }
    encode() {
        return {
            type: "constant",
            data: { delay: this.delay, maxRetries: this.maxRetries },
        };
    }
}
exports.Constant = Constant;
class Exponential {
    static type = "exponential";
    delay;
    maxRetries;
    factor;
    maxDelay;
    constructor({ delay = 1000, factor = 2, maxRetries = Number.MAX_SAFE_INTEGER, maxDelay = 30000, } = {}) {
        this.delay = delay;
        this.factor = factor;
        this.maxRetries = maxRetries;
        this.maxDelay = maxDelay;
    }
    next(attempt) {
        if (attempt < 0) {
            throw new Error("attempt must be greater than or equal to 0");
        }
        if (attempt > this.maxRetries) {
            return null;
        }
        if (attempt === 0) {
            return 0;
        }
        return Math.min(this.delay * this.factor ** attempt, this.maxDelay);
    }
    encode() {
        return {
            type: "exponential",
            data: { delay: this.delay, factor: this.factor, maxRetries: this.maxRetries, maxDelay: this.maxDelay },
        };
    }
}
exports.Exponential = Exponential;
class Linear {
    static type = "linear";
    delay;
    maxRetries;
    constructor({ delay = 1000, maxRetries = Number.MAX_SAFE_INTEGER } = {}) {
        this.delay = delay;
        this.maxRetries = maxRetries;
    }
    next(attempt) {
        if (attempt < 0) {
            throw new Error("attempt must be greater than or equal to 0");
        }
        if (attempt > this.maxRetries) {
            return null;
        }
        return this.delay * attempt;
    }
    encode() {
        return {
            type: "linear",
            data: { delay: this.delay, maxRetries: this.maxRetries },
        };
    }
}
exports.Linear = Linear;
class Never {
    static type = "never";
    next(attempt) {
        if (attempt < 0) {
            throw new Error("attempt must be greater than or equal to 0");
        }
        return attempt === 0 ? 0 : null;
    }
    encode() {
        return {
            type: "never",
            data: {},
        };
    }
}
exports.Never = Never;
