// @resonatehq/sdk/dist/src/tasks.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.Tasks = void 0;
const network_1 = require("../dev/network");
class Tasks {
    network;
    constructor(network = new network_1.LocalNetwork()) {
        this.network = network;
    }
    claim(id, counter, pid, ttl) {
        return new Promise((resolve, reject) => {
            this.network.send({
                kind: "claimTask",
                id: id,
                counter: counter,
                processId: pid,
                ttl: ttl,
            }, (err, res) => {
                if (err) {
                    // TODO: reject with more information
                    reject(Error("not implemented"));
                    return;
                }
                resolve(res.message);
            });
        });
    }
    complete(id, counter) {
        return new Promise((resolve, reject) => {
            this.network.send({
                kind: "completeTask",
                id: id,
                counter: counter,
            }, (err, res) => {
                if (err) {
                    // TODO: reject with more information
                    reject(Error("not implemented"));
                    return;
                }
                resolve(res.task);
            });
        });
    }
    heartbeat(pid) {
        return new Promise((resolve, reject) => {
            this.network.send({
                kind: "heartbeatTasks",
                processId: pid,
            }, (err, res) => {
                if (err) {
                    // TODO: reject with more information
                    reject(Error("not implemented"));
                    return;
                }
                resolve(res.tasksAffected);
            });
        });
    }
}
exports.Tasks = Tasks;
