// @resonatehq/sdk/dist/sim/src/worker.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || (function () {
    var ownKeys = function(o) {
        ownKeys = Object.getOwnPropertyNames || function (o) {
            var ar = [];
            for (var k in o) if (Object.prototype.hasOwnProperty.call(o, k)) ar[ar.length] = k;
            return ar;
        };
        return ownKeys(o);
    };
    return function (mod) {
        if (mod && mod.__esModule) return mod;
        var result = {};
        if (mod != null) for (var k = ownKeys(mod), i = 0; i < k.length; i++) if (k[i] !== "default") __createBinding(result, mod, k[i]);
        __setModuleDefault(result, mod);
        return result;
    };
})();
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.WorkerProcess = void 0;
const encryptor_1 = require("../../src/encryptor");
const exceptions_1 = __importDefault(require("../../src/exceptions"));
const handler_1 = require("../../src/handler");
const heartbeat_1 = require("../../src/heartbeat");
const options_1 = require("../../src/options");
const resonate_inner_1 = require("../../src/resonate-inner");
const tracer_1 = require("../../src/tracer");
const util = __importStar(require("../../src/util"));
const simulator_1 = require("./simulator");
class SimulatedMessageSource {
    maybeCorruptData;
    pid;
    group;
    unicast;
    anycast;
    subscriptions = { invoke: [], resume: [], notify: [] };
    constructor(iaddr, gaddr, maybeCorruptData) {
        this.maybeCorruptData = maybeCorruptData;
        this.pid = iaddr;
        this.group = gaddr;
        this.unicast = `sim://uni@${gaddr}/${iaddr}`;
        this.anycast = `sim://any@${gaddr}/${iaddr}`;
    }
    start() { }
    recv(msg) {
        for (const callback of this.subscriptions[msg.type]) {
            callback(this.maybeCorruptData(msg));
        }
    }
    subscribe(type, callback) {
        this.subscriptions[type].push(callback);
    }
    stop() { }
    match(target) {
        return `sim://any@${target}`;
    }
}
class SimulatedNetwork {
    source;
    target;
    correlationId = 1;
    currentTime = 0;
    prng;
    deliveryOptions;
    buffer = [];
    callbacks = {};
    messageSource;
    constructor(iaddr, gaddr, prng, { charFlipProb = 0 }, source, target) {
        this.source = source;
        this.target = target;
        this.prng = prng;
        this.deliveryOptions = { charFlipProb };
        this.messageSource = new SimulatedMessageSource(iaddr, gaddr, (data) => this.maybeCorruptData(data));
    }
    start() { }
    getMessageSource() {
        return this.messageSource;
    }
    send(req, cb) {
        const message = new simulator_1.Message(this.source, this.target, req, {
            requ: true,
            correlationId: this.correlationId++,
        });
        const callback = (err, res) => {
            if (res !== undefined) {
                util.assert(res.kind === req.kind, "res kind must match req kind");
                cb(undefined, res);
            }
            else {
                cb(err);
            }
        };
        this.callbacks[message.head.correlationId] = { callback, timeout: this.currentTime + 50000 };
        this.buffer.push(message);
    }
    stop() { }
    time(time) {
        this.currentTime = time;
        // Then, check for timed-out callbacks
        for (const key in this.callbacks) {
            const cb = this.callbacks[key];
            const hasTimedOut = cb.timeout < this.currentTime;
            if (hasTimedOut) {
                cb.callback(exceptions_1.default.SERVER_ERROR("Request timed out", true));
                delete this.callbacks[key];
            }
        }
    }
    process(message) {
        if (message.isResponse()) {
            util.assert(message.source === this.target);
            util.assert(message.target === this.source);
            const correlationId = message.head?.correlationId;
            const entry = correlationId && this.callbacks[correlationId];
            if (entry) {
                const msg = message;
                if (msg.data.err) {
                    util.assert(msg.data.res === undefined);
                    entry.callback(msg.data.err);
                }
                else {
                    util.assertDefined(msg.data.res);
                    entry.callback(undefined, this.maybeCorruptData(msg.data.res));
                }
                delete this.callbacks[correlationId];
            }
        }
        else {
            util.assert(message.source.kind === this.target.kind && message.source.iaddr === this.target.iaddr);
            this.messageSource.recv(message.data);
        }
    }
    flush() {
        // Finally, flush the buffer
        const flushed = this.buffer;
        this.buffer = [];
        return flushed;
    }
    maybeCorruptData(data) {
        // Serialize the data to a string
        let jsonStr = JSON.stringify(data);
        // Randomly decide whether to corrupt
        const shouldCorrupt = this.prng.next() < this.deliveryOptions.charFlipProb;
        if (!shouldCorrupt)
            return data;
        // Pick a random index to corrupt
        const idx = Math.floor(this.prng.next() * jsonStr.length);
        // Corrupt that character
        jsonStr = `${jsonStr.slice(0, idx)}X${jsonStr.slice(idx + 1)}`;
        // Return corrupted string, even if it's invalid JSON
        return jsonStr;
    }
}
class WorkerProcess extends simulator_1.Process {
    iaddr;
    gaddr;
    clock;
    network;
    registry;
    resonate;
    constructor(prng, clock, encoder, registry, { charFlipProb = 0 }, iaddr, gaddr) {
        super(iaddr, gaddr);
        this.iaddr = iaddr;
        this.gaddr = gaddr;
        this.clock = clock;
        this.network = new SimulatedNetwork(iaddr, gaddr, prng, { charFlipProb: 0 }, (0, simulator_1.unicast)(iaddr), (0, simulator_1.unicast)("server"));
        this.registry = registry;
        const messageSource = this.network.getMessageSource();
        this.resonate = new resonate_inner_1.ResonateInner({
            unicast: messageSource.unicast,
            anycast: messageSource.anycast,
            pid: iaddr,
            ttl: 5000,
            clock: this.clock,
            network: this.network,
            handler: new handler_1.Handler(this.network, encoder, new encryptor_1.NoopEncryptor()),
            messageSource: this.network.getMessageSource(),
            registry: registry,
            heartbeat: new heartbeat_1.NoopHeartbeat(),
            dependencies: new Map(),
            optsBuilder: new options_1.OptionsBuilder({ match: messageSource.match, idPrefix: "" }),
            verbose: false,
            tracer: new tracer_1.NoopTracer(),
        });
    }
    tick(tick, messages) {
        this.log(tick, "[recv]", messages);
        this.network.time(this.clock.time);
        for (const message of messages) {
            this.network.process(message);
        }
        const responses = this.network.flush();
        this.log(tick, "[send]", responses);
        return responses;
    }
}
exports.WorkerProcess = WorkerProcess;
